### PR TITLE
CUMULUS-2939: Add StartAsyncOperation lambda

### PR DESCRIPTION
**Pull request by jennyhliu**
**Summary:** Summary of changes

Addresses [CUMULUS-2939: Async operation (bulkDelete) endpoint timeout](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2939)

## Changes

* Added `@cumulus/api/lambdas/start-async-operation` to start an async operation
* Updated `@cumulus/api` `granules/bulk*`, `elasticsearch/index-from-database` and
    `POST reconciliationReports` endpoints to invoke StartAsyncOperation lambda
* cumulus-api document update: https://github.com/nasa/cumulus-api/pull/320

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [x] Integration tests

### CODE REVIEW

File: packages/api/endpoints/elasticsearch.js [hunk 1]
```diff
@@ -1,17 +1,16 @@
...
-const asyncOperations = require('@cumulus/async-operations');
 const { IndexExistsError } = require('@cumulus/errors');
 const { defaultIndexAlias, Search } = require('@cumulus/es-client/search');
 const { createIndex } = require('@cumulus/es-client/indexer');
 
 const { asyncOperationEndpointErrorHandler } = require('../app/middleware');
-
-const models = require('../models');
 const { getFunctionNameFromRequestContext } = require('../lib/request');
+const startAsyncOperation = require('../lib/startAsyncOperation');
```
Comment (Jkovarik) [inline packages/api/endpoints/elasticsearch.js:13]: Can we generally import this like const { invokeStartAsyncOperationLambda } = require('../lib/startAsyncOperation') and avoid all of the extra text/typing of startAsyncOperation.invokeStartAsyncOperationLambda?

**Review by Jkovarik — COMMENTED** [review]
Pushing review work in progress as I'd started a review instead of pushing comments as they were made 😅

File: packages/api/lib/startAsyncOperation.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+'use strict';
+
+const { lambda } = require('@cumulus/aws-client/services');
+
+async function invokeStartAsyncOperationLambda(event) {
+  if (!process.env.StartAsyncOperationLambda) {
+    throw new Error('The StartAsyncOperationLambda environment variable is not set.');
+  }
+
+  await lambda().invoke({
```
[the remark this answers is not in the document]
Comment (jennyhliu) [reply packages/api/lib/startAsyncOperation.js:10]: We invoke a function asynchronously by setting `InvocationType to Event`.  The default is `RequestResponse`

[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/api/lib/startAsyncOperation.js:10]: 🤦🏻     I knew I was missing something obvious.   Thanks :)

File: packages/api/tests/endpoints/test-reconciliation-reports.js [hunk 3]
```diff
@@ -423,25 +394,8 @@ test.serial('POST returns a 500 when failing to create an async operation', asyn
   }
 });
 
-test.serial('POST returns a 503 when there is an ecs error', async (t) => {
```
Comment (Jkovarik) [inline packages/api/tests/endpoints/test-reconciliation-reports.js:426]: Do we still need some way to handle providing this feedback to users?  Else it seems like in these failure cases the lambda will execute and *nothing* will show up in the DB *or* the API response.     

 In theory we could be updating the record to fail with information if `startAsyncOperation` throws, at the cost of adding database configuration to the invocation lambda.

**Review by Jkovarik — APPROVED** [review]
Jenny - thanks for your patience with this review!     This is really great, I really like the ID assignment/lambda invocation approach and it shouldn't impact lambda limits too much given this shouldn't *generally* be being scaled horizontally.   👍🏻  

Just a couple of minor comments for your consideration, and I think we need to consider addressing the issue with invocation failure and this should be good to go.

**Review by Jkovarik — CHANGES_REQUESTED** [review]
Changing to change request (the original comment should have set this)

File: packages/api/tests/endpoints/test-reconciliation-reports.js (hunk 3 above)
Comment (jennyhliu) [reply packages/api/tests/endpoints/test-reconciliation-reports.js:426]: Updated startAsyncOperation  method to create asyncOperation record when startEcsTask fails
[fde322f](https://github.com/nasa/cumulus/pull/2962/commits/fde322ff66076714738950457eb5e7986a78c067)

File: packages/async-operations/src/async_operations.ts [hunk 4]
```diff
@@ -230,7 +234,7 @@ export const startAsyncOperation = async (
     throw new MissingRequiredArgument(`callerLambdaName must be specified to start new async operation, received: ${callerLambdaName}`);
   }
 
-  const id = uuidv4();
+  const id = params.asyncOperationId ?? uuidv4();
   // Store the payload to S3
```
[the remark this answers is not in the document]
Comment (jennyhliu) [reply packages/async-operations/src/async_operations.ts:238]: added unit test to verify the asyncOperationId is used if passed in

File: packages/api/tests/endpoints/test-reconciliation-reports.js (hunk 3 above)
Comment (jennyhliu) [reply packages/api/tests/endpoints/test-reconciliation-reports.js:426]: Tested manually by updating launchType to 'FARGATE2' in the code, record is saved when startEcsTask fails.
```
{
 "id": "cd27a334-5451-477d-88d6-c35c1b1e24cc",
 "createdAt": 1653505934999,
 "description": "Bulk granule reingest run on 2 granules",
 "operationType": "Bulk Granule Reingest",
 "output": "{\"name\":\"InvalidParameterException\",\"message\":\"launch type should be one of [EC2,FARGATE,EXTERNAL]\",\"stack\":\"InvalidParameterException: launch type should be one of [EC2,FARGATE,EXTERNAL]\\n    at Request.extractError (/var/runtime/node_modules/aws-sdk/lib/protocol/json.js:52:27)\\n    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:106:20)\\n    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:78:10)\\n    at Request.emit (/var/runtime/node_modules/aws-sdk/lib/request.js:686:14)\\n    at Request.transition (/var/runtime/node_modules/aws-sdk/lib/request.js:22:10)\\n    at AcceptorStateMachine.runTo (/var/runtime/node_modules/aws-sdk/lib/state_machine.js:14:12)\\n    at /var/runtime/node_modules/aws-sdk/lib/state_machine.js:26:10\\n    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:38:9)\\n    at Request.<anonymous> (/var/runtime/node_modules/aws-sdk/lib/request.js:688:12)\\n    at Request.callListeners (/var/runtime/node_modules/aws-sdk/lib/sequential_executor.js:116:18)\"}",
 "status": "RUNNER_FAILED",
 "updatedAt": 1653505934999
}
```

File: tf-modules/archive/start_async_operation.tf [hunk 5]
```diff
@@ -0,0 +1,104 @@
+resource "aws_lambda_function" "start_async_operation" {
+  function_name    = "${var.prefix}-StartAsyncOperation"
+  filename         = "${path.module}/../../packages/api/dist/startAsyncOperation/lambda.zip"
+  source_code_hash = filebase64sha256("${path.module}/../../packages/api/dist/startAsyncOperation/lambda.zip")
+  handler          = "index.handler"
+  role             = aws_iam_role.start_async_operation.arn
+  runtime          = "nodejs14.x"
+  timeout          = 300
+  memory_size      = 256
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply tf-modules/archive/start_async_operation.tf]: 80e5b753c618dc42c3c0df2966595eadbb36431

[the remark this answers is not in the document]
Comment (Jkovarik) [reply tf-modules/archive/start_async_operation.tf]: https://github.com/nasa/cumulus/commit/a80e5b753c618dc42c3c0df2966595eadbb36431

File: packages/async-operations/src/async_operations.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/async-operations/src/async_operations.ts:238]: https://github.com/nasa/cumulus/commit/fde322ff66076714738950457eb5e7986a78c067

[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/async-operations/src/async_operations.ts:238]: https://github.com/nasa/cumulus/commit/fde322ff66076714738950457eb5e7986a78c067

File: tf-modules/archive/start_async_operation.tf [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+      "ec2:DeleteNetworkInterface",
+      "ec2:DescribeNetworkInterfaces",
+      "lambda:GetFunctionConfiguration",
+      "lambda:invokeFunction",
+      "logs:CreateLogGroup",
+      "logs:CreateLogStream",
+      "logs:DescribeLogStreams",
+      "logs:PutLogEvents",
+    ]
+    resources = ["*"]
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply tf-modules/archive/start_async_operation.tf]: Thanks for renaming the file and that update.

File: packages/api/tests/endpoints/test-reconciliation-reports.js (hunk 3 above)
Comment (Jkovarik) [reply packages/api/tests/endpoints/test-reconciliation-reports.js:426]: Awesome!  Thanks for making that update!!

**Review by Jkovarik — APPROVED** [review]
Thanks for addressing those comments @jennyhliu , this looks fantastic!

**Pull request merged**